cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/query/IndexUpdaterTest.java

```diff
@@ -125,6 +127,123 @@ public void updateOneFieldIndex() throws Exception {
         }
     }
 
+
+    @Test
+    public void updateOneFieldIndexMultithreaded() throws Exception {
+        createIndex("basic", Arrays.<Object>asList("name"));
+
+        assertThat(getIndexSequenceNumber("basic"), is(0l));
+
+        String table = IndexManager.tableNameForIndex("basic");
+        String sql = String.format("SELECT * FROM %s", table);
+        Cursor cursor = null;
+        try {
+            SQLDatabase db = TestUtils.getDatabaseConnectionToExistingDb(this.db);
+            cursor = db.rawQuery(sql, new String[]{});
+            assertThat(cursor.getCount(), is(0));
+        } catch (SQLException e) {
+            Assert.fail(String.format("SQLException occurred executing %s: %s", sql, e));
+        } finally {
+            DatabaseUtils.closeCursorQuietly(cursor);
+        }
+
+        final int nDocs = 50;
+        final int nThreads = 8;
+        final int nUpdates = 10; // update index after how many docs?
+
+        // populate nDocs documents per thread, across nThreads simultaneously
+        // and update index after each insert
+
+        final CountDownLatch latch = new CountDownLatch(nThreads);
+
+        class PopulateThread extends Thread {
+            @Override
+            public void run() {
+                try {
+                    latch.await();
+                } catch (InterruptedException e) {
+                    Assert.fail(e.toString());
+                }
+                for (int i = 0; i < nDocs; i++) {
+                    MutableDocumentRevision rev = new MutableDocumentRevision();
+                    rev.docId = String.format("id_%d_%s",i,Thread.currentThread().getName());
+                    System.out.println("creating "+rev.docId);
+                    Map<String, Object> bodyMap = new HashMap<String, Object>();
+                    if (i % 2 == 0) {
+                        bodyMap.put("name", "mike");
+                    } else {
+                        bodyMap.put("name", "tom");
+                    }
+                    rev.body = DocumentBodyFactory.create(bodyMap);
+                    try {
+                        ds.createDocumentFromRevision(rev);
+                    } catch (DocumentException de) {
+                        Assert.fail("Exception thrown when creating revision " + de);
+                    }
+                    // batch up index updates
+                    if (i % nUpdates == 0) {
+                        IndexUpdater.updateIndex("basic", fields, db, ds, im.getQueue());
+                    }
+                }
+            }
+        }
+
+        List<Thread> threads = new ArrayList<Thread>();
+
+        // Create, start and wait for the threads to complete
+        for (int i = 0; i < nThreads; i++) {
+            threads.add(new PopulateThread());
+        }
+        for (Thread t : threads) {
+            t.start();
+            // try to get all threads to start simultaneously
+            latch.countDown();
```

Comment:
I think moving this inside the `run()` method gives a slightly tighter window of the threads all being notified at the same time because the last thread will be guaranteed to be in its `run` method before the last `countdown` happens. As we discussed earlier all of this is going to be very subject to thread scheduling anyway so it probably doesn't make much difference.